Feature request - Create stubs for public methods #11

rCartoux opened the pull request:
- Generate test methods for each public method available in the selected class.
- The test methods will be placed where the new token "$TestMethods$" is specified in the file template
- New templates are available : 
         - Tested object creation : code to instantiate the object to test in each test method
         - Tested object reference : name of the local variable or field of the object to test in the test method
- Added new test cases

rCartoux committed:
Debugging support, Microsoft.VisualStudio.ComponentModelHost upgrade

- Added start parameters in order to debug the extension easily by opening another instance of visual studio
- restored nuget for Microsoft.VisualStudio.ComponenModelHost

rCartoux committed:
First version Test methods generation, calling the public method under test with default parameters, and assert Fail. No template.

rCartoux committed:
Handle methods having the same name and different signatures

rCartoux committed:
Fixed test method generation for async methods : adjust test method modifier and return value, await the call to the method under test, add using statement if needed.

rCartoux committed:
Added templates for creation and reference of the tested object : settings, ui.

rCartoux committed:
Make test cases green again

Adjusted expectations by removing the default test method.

rCartoux committed:
Added test cases, bug fixes

- Added test cases for a class with methods
- Fixed default tested object creation for AutoMoq
- Fixed missing semicolumn in test method call without parameter
- Fixed handling of awaitable methods in case there is no async modifier

rCartoux committed:
Refresh test method templates when changing test or mock framework

Fixed ui issue where the tested object creation and reference template where not refreshed after changing the test framework or the mock framework.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/UnitTestBoilerplateGenerator?pullRequest=28) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/UnitTestBoilerplateGenerator?pullRequest=28) before we can accept your contribution.<br/><br/>:x: rCartoux [sign now](https://cla.opensource.microsoft.com/Microsoft/UnitTestBoilerplateGenerator?pullRequest=28)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/UnitTestBoilerplateGenerator?pullRequest=28) it.</sub>

RandomEngy commented on src/Services/TestGenerationService.cs:
```diff
@@ -516,7 +625,104 @@ private void WriteExplicitConstructor(StringBuilder builder, TestGenerationConte
+
+					for (int i = 0; i < numberOfParameters; i++)
+					{
+						builder.Append($"	default({methodDescriptor.MethodParameterNames[i]})");
```
Couple of thoughts here:

1. You should bring in namespaces for types you include here.
2. It might be better to just put "TODO" in here since that will require that the writer of the test think about the input needed rather than relying on default behavior.

RandomEngy commented on src/Services/TestGenerationService.cs:
```diff
@@ -516,7 +625,104 @@ private void WriteExplicitConstructor(StringBuilder builder, TestGenerationConte
+				string asyncModifier = methodDescriptor.IsAsync ? "async" : string.Empty;
+
+				builder.AppendLine($"[{context.TestFramework.TestMethodAttribute}]");
+				builder.AppendLine($"public {asyncModifier} {returnType} {testMethodPrefix}_Condition_Expectation()");
```
Different developers have different ways they like to name methods. I've been doing them WhenXAndYSoAndSo_ThenOutputIsFalse. This would probably be a good place for another format to accommodate different styles. In this case it would be just a test method identifier and not a "prefix".

RandomEngy commented on src/Services/TestGenerationService.cs:
```diff
@@ -516,7 +625,104 @@ private void WriteExplicitConstructor(StringBuilder builder, TestGenerationConte
+					builder.Append("await ");
+				}
+
+				builder.Append($"{testedObjectReference}.{methodDescriptor.Name}(");
```
If the method returns something; it should probably be put into a variable so the test writer can easily check it in the Assert section.

RandomEngy commented on src/Services/TestGenerationService.cs:
```diff
@@ -516,7 +625,104 @@ private void WriteExplicitConstructor(StringBuilder builder, TestGenerationConte
+
+			var testMethodPrefixes = new List<string>();
+
+			int idx = 0;
```
I think you'd only need a bool isFirstMethod, since that index is not used elsewhere.

RandomEngy commented on src/Services/TestGenerationService.cs:
```diff
@@ -516,7 +625,104 @@ private void WriteExplicitConstructor(StringBuilder builder, TestGenerationConte
+
+			int idx = 0;
+
+			foreach (var methodDescriptor in context.MethodDeclarations)
```
If there are no public methods it would be better to have a single "generic" test case like with the old code than to have nothing at all. You could have the tested object creation there but no "actions".

rCartoux commented:
Thanks!
I am thinking now about some problems I didn't see:
- When the class has methods with arguments having ref or out modifiers : does not work now, need to add local variables. Also need to check what happens if there is params keyword in the arguments
- Now content tokens are accepted and replaced in the tested object creation and reference template, which does not make sense except for $ExplicitConstructor$ in the creation one. But maybe it is a separate issue about validation in general (define and validate in the settings ui which tokens are allowed in which templates) ?

RandomEngy reviewed (changes requested):
I like how you fixed up the test cases and added the new format fields with conditional visibility. I think it will be a great addition to the extension. I have just a few comments on how to round out the feature.

rCartoux commented:
Ok thanks! will do

RandomEngy commented:
You're right, the "accepted tokens" list might need a bit of shuffling around. Also https://github.com/Microsoft/UnitTestBoilerplateGenerator/wiki/Custom-Format-Tokens will need an update.

rCartoux committed:
Removed unnecessary index variable

rCartoux committed:
Store the result of the tested method call for user assertion

- Create local variable in the test method
- Only if the tested method returns void or Task (not generic)

rCartoux committed:
Added format for test method name and token for the name of the method under test

- added format field "Test method name format"
- added token for the name of the method under test, to be used only within the test method name format

rCartoux committed:
Refresh test method name format when changing mock or test framework

rCartoux committed:
Write defaul test method if the tested class has no public methods

RandomEngy commented:
With your latest version I'm getting an exception when trying to generate from ClassWithMethods to VSTestCases:

---------------------------
System.NotSupportedException: Parameter type not supported

   at UnitTestBoilerplate.Services.TestGenerationService.GetArgumentDescriptors(List`1 argumentList, SemanticModel semanticModel, MockFramework mockFramework) in D:\git\UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line 324
   at UnitTestBoilerplate.Services.TestGenerationService.<CollectTestGenerationContextAsync>d__9.MoveNext() in D:\git\UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line 238
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at UnitTestBoilerplate.Services.TestGenerationService.<CollectTestGenerationContextAsync>d__15.MoveNext() in D:\git\UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line 343
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at UnitTestBoilerplate.Services.TestGenerationService.<GenerateUnitTestFileAsync>d__6.MoveNext() in D:\git\UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line 57
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at UnitTestBoilerplate.ViewModel.CreateUnitTestBoilerplateViewModel.<CreateUnitTestAsync>d__54.MoveNext() in D:\git\UnitTestBoilerplateGenerator\src\ViewModel\CreateUnitTestBoilerplateViewModel.cs:line 181
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at UnitTestBoilerplate.ViewModel.CreateUnitTestBoilerplateViewModel.<<get_CreateUnitTestCommand>b__53_0>d.MoveNext() in D:\git\UnitTestBoilerplateGenerator\src\ViewModel\CreateUnitTestBoilerplateViewModel.cs:line 166

rCartoux commented:
Hi David,
yes I just saw thx, I have a look

Robin

On 30 June 2018 at 14:06, David Rickard <notifications@github.com> wrote:

> With your latest version I'm getting an exception when trying to generate
> from ClassWithMethods to VSTestCases:
> ------------------------------
>
> System.NotSupportedException: Parameter type not supported
>
> at UnitTestBoilerplate.Services.TestGenerationService.
> GetArgumentDescriptors(List1 argumentList, SemanticModel semanticModel,
> MockFramework mockFramework) in D:\git\UnitTestBoilerplateGenerator\
> src\Services\TestGenerationService.cs:line 324 at
> UnitTestBoilerplate.Services.TestGenerationService.<
> CollectTestGenerationContextAsync>d__9.MoveNext() in D:\git\
> UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line
> 238 --- End of stack trace from previous location where exception was
> thrown --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task
> task) at System.Runtime.CompilerServices.TaskAwaiter.
> HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.
> CompilerServices.TaskAwaiter1.GetResult()
> at UnitTestBoilerplate.Services.TestGenerationService.d__15.MoveNext() in
> D:\git\UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line
> 343
> --- End of stack trace from previous location where exception was thrown
> ---
> at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task
> task)
> at System.Runtime.CompilerServices.TaskAwaiter.
> HandleNonSuccessAndDebuggerNotification(Task task)
> at System.Runtime.CompilerServices.TaskAwaiter1.GetResult() at
> UnitTestBoilerplate.Services.TestGenerationService.<
> GenerateUnitTestFileAsync>d__6.MoveNext() in D:\git\
> UnitTestBoilerplateGenerator\src\Services\TestGenerationService.cs:line
> 57 --- End of stack trace from previous location where exception was thrown
> --- at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task
> task) at System.Runtime.CompilerServices.TaskAwaiter.
> HandleNonSuccessAndDebuggerNotification(Task task) at System.Runtime.
> CompilerServices.TaskAwaiter1.GetResult()
> at UnitTestBoilerplate.ViewModel.CreateUnitTestBoilerplateViewModel.d__54.MoveNext()
> in D:\git\UnitTestBoilerplateGenerator\src\ViewModel\
> CreateUnitTestBoilerplateViewModel.cs:line 181
> --- End of stack trace from previous location where exception was thrown
> ---
> at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task
> task)
> at System.Runtime.CompilerServices.TaskAwaiter.
> HandleNonSuccessAndDebuggerNotification(Task task)
> at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
> at UnitTestBoilerplate.ViewModel.CreateUnitTestBoilerplateViewModel.<<get_
> CreateUnitTestCommand>b__53_0>d.MoveNext() in D:\git\
> UnitTestBoilerplateGenerator\src\ViewModel\CreateUnitTestBoilerplateViewModel.cs:line
> 166
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/UnitTestBoilerplateGenerator/pull/28#issuecomment-401540145>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV8YZF3oy4D7PuGezXKovuZsY73VuaJ7ks5uB3fqgaJpZM4U1IK9>
> .
>

rCartoux committed:
Create local variables for method arguments with type information and assigned to TODO

- Remove default() behavior
- create local variable for each method argument

RandomEngy commented:
Awesome! Can you re-run the self-test to update the expected output for the default method changes you made? And let me know when you're all done with the changes and I'll ship it.

rCartoux committed:
Write usings for namespaces needed by method arguments

rCartoux commented:
Ok!

Unfortunately there are still some cases I need to check like arguments
which are one or multidimensional arrays, params/ref/out modifiers etc

Then I ll do the test cases, maybe add some new ones. I let you know when
its ready

Thx for the feedback

Cheers

On 30 June 2018 at 14:48, David Rickard <notifications@github.com> wrote:

> Awesome! Can you re-run the self-test to update the expected output for
> the default method changes you made? And let me know when you're all done
> with the changes and I'll ship it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/UnitTestBoilerplateGenerator/pull/28#issuecomment-401542393>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AV8YZC1DCHppUQobsJpET7fbLOWl_WmUks5uB4GfgaJpZM4U1IK9>
> .
>

rCartoux committed:
Handle identified arguments like DateTime

rCartoux committed:
Handle array parameters

rCartoux committed:
Fixed declaration of variables for generic type arguments

rCartoux committed:
Handle parameter modifiers out, ref

rCartoux commented:
should be ok now

rCartoux committed:
Fix VS crash if test methods content token is written in test methods formats

- For tested object creation, only allow class name tokens and explicit constructor to be replaced
- For tested object reference,  only allow class name tokens to be replaced
- Tested method name format only allows method identifier token to be replaced

rCartoux committed:
Fix missing reset of test method name format

rCartoux committed:
Make test cases pass

Added method test cases in ClassWithMethods

RandomEngy commented:
Now getting this when I invoke the create unit test boilerplate option:

```
System.Reflection.TargetParameterCountException
  HResult=0x8002000E
  Message=Parameter count mismatch.
  Source=mscorlib
  StackTrace:
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.VisualStudio.Composition.ReflectionHelpers.Instantiate(MethodBase ctorOrFactoryMethod, Object[] arguments)
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass53_0.<CreateExport>b__0()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at System.Lazy`1.get_Value()
   at Microsoft.VisualStudio.Composition.NetFxAdapters.MefV1ExportProvider.<>c__DisplayClass12_0.<UnwrapExport>b__1()
   at System.ComponentModel.Composition.Primitives.Export.GetExportedValueCore()
   at System.ComponentModel.Composition.Primitives.Export.get_Value()
   at System.ComponentModel.Composition.ReflectionModel.ImportingItem.Cast(Type type, Export export)
   at System.ComponentModel.Composition.ReflectionModel.ImportingItem.CastSingleExportToImportType(Type type, Export export)
   at System.ComponentModel.Composition.ReflectionModel.ImportingItem.CastExportsToSingleImportType(Export[] exports)
   at System.ComponentModel.Composition.ReflectionModel.ImportingItem.CastExportsToImportType(Export[] exports)
   at System.ComponentModel.Composition.ReflectionModel.ReflectionComposablePart.SetImport(ImportingItem item, Export[] exports)
   at System.ComponentModel.Composition.ReflectionModel.ReflectionComposablePart.SetImport(ImportDefinition definition, IEnumerable`1 exports)
   at System.ComponentModel.Composition.Hosting.ImportEngine.PartManager.TrySetImport(ImportDefinition import, Export[] exports)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportSubset(PartManager partManager, IEnumerable`1 imports, AtomicComposition atomicComposition)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImportsStateMachine(PartManager partManager, ComposablePart part)
   at System.ComponentModel.Composition.Hosting.ImportEngine.TrySatisfyImports(PartManager partManager, ComposablePart part, Boolean shouldTrackImports)
   at System.ComponentModel.Composition.Hosting.ImportEngine.SatisfyImportsOnce(ComposablePart part)
   at System.ComponentModel.Composition.Hosting.CompositionContainer.SatisfyImportsOnce(ComposablePart part)
   at Microsoft.VisualStudio.Composition.NetFxAdapters.CompositionService.SatisfyImportsOnce(ComposablePart part)
   at System.ComponentModel.Composition.AttributedModelServices.SatisfyImportsOnce(ICompositionService compositionService, Object attributedPart)
   at UnitTestBoilerplate.View.CreateUnitTestBoilerplateDialog..ctor() in D:\git\UnitTestBoilerplateGenerator\src\View\CreateUnitTestBoilerplateDialog.xaml.cs:line 20
   at UnitTestBoilerplate.Commands.CreateUnitTestBoilerplateCommand.MenuItemCallback(Object sender, EventArgs e) in D:\git\UnitTestBoilerplateGenerator\src\Commands\CreateUnitTestBoilerplateCommand.cs:line 109
   at Microsoft.VisualStudio.Shell.OleMenuCommand.Invoke(Object inArg, IntPtr outArg, OLECMDEXECOPT options)
   at Microsoft.VisualStudio.Shell.OleMenuCommandService.Microsoft.VisualStudio.OLE.Interop.IOleCommandTarget.Exec(Guid& commandGroup, UInt32 nCmdId, UInt32 nCmdExcept, IntPtr pIn, IntPtr vOut)
   at Microsoft.VisualStudio.Shell.Package.Microsoft.VisualStudio.OLE.Interop.IOleCommandTarget.Exec(Guid& guidGroup, UInt32 nCmdId, UInt32 nCmdExcept, IntPtr pIn, IntPtr vOut)

```

RandomEngy commented:
Thanks again. Released with 1.9. Would you be able to also update the "tokens" wiki page? https://github.com/Microsoft/UnitTestBoilerplateGenerator/wiki/Custom-Format-Tokens

rCartoux commented:
sure!

RandomEngy commented:
Also a super easy way to help is rate it on the marketplace: https://marketplace.visualstudio.com/items?itemName=RandomEngy.UnitTestBoilerplateGenerator it only has a handful of reviews. There is a guy who hasn't updated his 1-star review months after I added xUnit support for him. :/

rCartoux commented:
I have updated the custom format tokens page in my wiki fork : new TestMethods and TestedMethodName formats 
https://github.com/rCartoux/UnitTestBoilerplateGenerator-wiki

(not sure how to proceed with wiki issues / pull requests)

RandomEngy commented:
I'll just copy/paste. Thanks!
